Grid/qcd/utils/BaryonUtils.h

```diff
@@ -0,0 +1,244 @@
+ 
+ Source file: ./lib/qcd/utils/BaryonUtils.h
+ 
+ Copyright (C) 2019
+ 
+ Author: Felix Erben <felix.erben@ed.ac.uk>
+
+ This program is free software; you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; either version 2 of the License, or
+ (at your option) any later version.
+ 
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+ 
+ You should have received a copy of the GNU General Public License along
+ with this program; if not, write to the Free Software Foundation, Inc.,
+ 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ 
+ See the full license in the file "LICENSE" in the top level distribution directory
+ *************************************************************************************/
+/*  END LEGAL */
+#pragma once
+//#include <Grid/Hadrons/Global.hpp>
+#include <Grid/Eigen/unsupported/CXX11/Tensor>
+
+NAMESPACE_BEGIN(Grid);
+
+#undef DELTA_F_EQ_2
+
+template <typename FImpl>
+class BaryonUtils 
+{
+public:
+  typedef typename FImpl::ComplexField ComplexField;
+  typedef typename FImpl::FermionField FermionField;
+  typedef typename FImpl::PropagatorField PropagatorField;
+
+  typedef typename FImpl::SitePropagator pobj;
+  typedef typename ComplexField::vector_object vobj;
+
+  private: 
+  template <class mobj, class robj>
+  static void baryon_site(const mobj &D1,
+				 const mobj &D2,
+				 const mobj &D3,
+				 const Gamma GammaA,
+				 const Gamma GammaB,
+				 const int parity,
+				 const std::vector<int> &wick_contractions,
+  				 robj &result);
+  public:
+  static void ContractBaryons(const PropagatorField &q1_src,
+				 const PropagatorField &q2_src,
+				 const PropagatorField &q3_src,
+				 const Gamma GammaA,
+				 const Gamma GammaB,
+				 const char * quarks_snk,
+				 const char * quarks_src,
+				 const int parity,
+				 ComplexField &baryon_corr);
+  template <class mobj, class robj>
+  static void ContractBaryons_Sliced(const mobj &D1,
+				 const mobj &D2,
+				 const mobj &D3,
+				 const Gamma GammaA,
+				 const Gamma GammaB,
+				 const char * quarks_snk,
+				 const char * quarks_src,
+				 const int parity,
+				 robj &result);
+};
+
+template <class FImpl>
+template <class mobj, class robj>
+void BaryonUtils<FImpl>::baryon_site(const mobj &D1,
+						 const mobj &D2,
+						 const mobj &D3,
+						 const Gamma GammaA,
+						 const Gamma GammaB,
+						 const int parity,
+						 const std::vector<int> &wick_contraction,
+						 robj &result)
+{
+
+  Gamma g4(Gamma::Algebra::GammaT); //needed for parity P_\pm = 0.5*(1 \pm \gamma_4)
+
+
+  static const int epsilon[6][3] = {{0,1,2},{1,2,0},{2,0,1},{0,2,1},{2,1,0},{1,0,2}};
+  static const Complex epsilon_sgn[6]= {1,1,1,-1,-1,-1};

```

Comment:
Don't duplicate the epsilon values and signs, make it a static member of the `BaryonUtils` class